### PR TITLE
Fix missing data handling

### PR DESCRIPTION
**Pull request by mrocklin**
This fixes a few of the xfailing tests.  Going through this commit-by-commit probably makes sense.

### CODE REVIEW

**Comment by mrocklin** [comment]
======= 356 passed, 13 skipped, 15 xfailed, 2 warnings in 99.26 seconds ========

File: geopandas/tests/test_pandas_methods.py [hunk 1]
```diff
@@ -169,16 +170,15 @@ def test_select_dtypes(df):
 # Missing values
 
 
-@pytest.mark.xfail
 def test_fillna():
     # this currently does not work (it seems to fill in the second coordinate
     # of the point
```
Comment (jorisvandenbossche) [inline geopandas/tests/test_pandas_methods.py]: maybe remove this comment as well?

File: geopandas/_block.py [hunk 2]
```diff
@@ -10,10 +10,11 @@
 
 
 class GeometryBlock(NonConsolidatableMixIn, Block):
-    """ implement a geometry block with uint pointers to C objects
-    as underlying data"""
+    """ Pandas Geometry block with pointers to C GEOS objects """
     __slots__ = ()
 
+    _concatenator = concat
```
Comment (jorisvandenbossche) [inline geopandas/_block.py]: I think this was handled in the other PR?

**Comment by mrocklin** [comment]
> I still have to look at the changes in the tests, but in general: I think we need to discuss a bit the 'missing values' concept in general. To what extent do we see an empty geometry collection as missing as well? Eg isna regards it as missing, but in your new dropna not.

I believe that in master None is used to represent missing values.  In geopandas-cython we use empty polygons for missing values.  I think that this PR reverts back to `None`.  However from your comment about `isna` it appears that this reversion was not complete.

**Comment by jorisvandenbossche** [comment]
Quick exploration of missing/empty values in both master as geopandas-cython: http://nbviewer.jupyter.org/gist/jorisvandenbossche/704c799d80a00aef4b1244b97282235a (should probably open a new issue, as there are some discussion points like: distinguish missing vs empty? how to handle both in 'missing value' operations, ...)

**Comment by jdmcbr** [comment]
@jorisvandenbossche Thanks for documenting the different scenarios! I think there's almost certainly value in creating a separate issue to discuss; it would be beneficial to get a wider range of opinions on this, and a discussion in an issue on the cython branch labeled as tests related seems unlikely to catch the eye of anyone wandering though.

File: geopandas/geoseries.py [hunk 3]
```diff
@@ -326,21 +326,23 @@ def notnull(self):
...
+        assert method is None and not inplace
         if value is None:
             value = BaseGeometry()
-        return super(GeoSeries, self).fillna(value=value, method=method,
-                                             inplace=inplace, **kwargs)
+        return GeoSeries(self._geometry_array.fillna(value), index=self.index,
+                         crs=self.crs)
+
+    def dropna(self, method=None, inplace=False, **kwargs):
+        """ Fill NA/NaN values with a geometry (empty polygon by default) """
```
Comment (jorisvandenbossche) [inline geopandas/geoseries.py]: This is the docstring of fillna

File: geopandas/geoseries.py [hunk 4]
```diff
@@ -326,21 +326,23 @@ def notnull(self):
...
+        assert method is None and not inplace
+        return GeoSeries(self._geometry_array[~self.isna()],
+                         index=self.index[~self.isna()],
+                         crs=self.crs)
```
Comment (jorisvandenbossche) [inline geopandas/geoseries.py]: name here as well

**Comment by mrocklin** [comment]
I've pushed up a fix, I think, for the NA handling.

File: geopandas/tests/test_geoseries.py [hunk 5]
```diff
@@ -102,9 +102,9 @@ def test_geom_equals_align(self):
 
     def test_align(self):
         a1, a2 = self.a1.align(self.a2)
-        assert a2['A'].is_empty
+        assert not a2['A']
```
Comment (jorisvandenbossche) [inline geopandas/tests/test_geoseries.py]: Can we do `is None` here instead of `not`? I personally find that more explicit, as now my first idea is of it checking for False

File: geopandas/vectorized.pyx [hunk 6]
```diff
@@ -1210,6 +1210,26 @@ class GeometryArray(object):
...
+
+        Value should be a BaseGeometry
+
+        Returns a copy
+        """
+        base = [self]
+        if isinstance(value, BaseGeometry):
+            base.append(value)
+            value = value.__geom__
+        if value is None:
```
Comment (jorisvandenbossche) [inline geopandas/vectorized.pyx]: if -> elif?

(and should we after this raise a ValueError that the passed value is not of correct type ?)

File: geopandas/_block.py (hunk 2 above)
Comment (jorisvandenbossche) [reply geopandas/_block.py]: this one can be removed

**Comment by mrocklin** [comment]
Handled the recent comments. Thanks for the review @jorisvandenbossche

File: geopandas/tests/test_vectorized.py [hunk 7]
```diff
@@ -412,3 +412,14 @@ def test_concat():
...
     assert len(c) == 5
+
+
+def test_fill():
+    p = Point(1, 2)
+    P2 = P.fill([0, 3], p)
+    assert P2[0].equals(p)
+    assert P2[3].equals(p)
+    with pytest.raises(TypeError) as info:
+        P.fill([1, 2], 123)
```
Comment (mrocklin) [inline geopandas/tests/test_vectorized.py]: This is failing with a TypeError despite the `pytest.raises` context manager.  I'm not sure why.  I suspect some Cython issue.  Do you have any experience with this @jorisvandenbossche ?

Comment (mrocklin) [reply geopandas/tests/test_vectorized.py]: Hrm, it looks like this doesn't fail on travis.  This is probably something to do with my local testing environment.

**Comment by mrocklin** [comment]
Any further thoughts here @jorisvandenbossche ?

File: geopandas/geoseries.py [hunk 8]
```diff
@@ -326,21 +326,26 @@ def notnull(self):
...
+        """ Fill NA/NaN values with a geometry (empty polygon by default) """
+        assert method is None and not inplace
         if value is None:
             value = BaseGeometry()
-        return super(GeoSeries, self).fillna(value=value, method=method,
-                                             inplace=inplace, **kwargs)
+        return GeoSeries(self._geometry_array.fillna(value), index=self.index,
+                         crs=self.crs, name=self.name)
+
+    def dropna(self, method=None, inplace=False, **kwargs):
```
Comment (jorisvandenbossche) [inline geopandas/geoseries.py]: I don't think `method` is keyword for dropna in pandas, so probably copied from fillna. Signature for dropna: ` def dropna(self, axis=0, inplace=False)`, but the axis cannot be anything else as 0

File: geopandas/tests/test_vectorized.py (hunk 7 above)
Comment (jorisvandenbossche) [reply geopandas/tests/test_vectorized.py]: Doesn't fail for me either

File: geopandas/tests/test_geoseries.py [hunk 9]
```diff
@@ -158,8 +158,7 @@ def test_transform(self):
     def test_fillna(self):
         # default is to fill with empty geometry
         na = self.na_none.fillna()
-        assert isinstance(na[2], BaseGeometry)
-        assert na[2].is_empty
+        assert not na[2]
```
Comment (jorisvandenbossche) [inline geopandas/tests/test_geoseries.py]: Why was this changed? Isn't it the default of fillna to fill with empty geometry? so shouldn't assert na[2].is_empty still work?

File: geopandas/geoseries.py (hunk 8 above)
Comment (mrocklin) [reply geopandas/geoseries.py]: Fixed.

However given that these aren't supported I am tempted to just not include them in the signature.  The default error message for an extra keyword seems decently informative.

What are your thoughts on this general practice?  My guess is that it is a deviation from normal geopandas development.  I'm happy to go either way.

File: geopandas/tests/test_geoseries.py (hunk 9 above)
Comment (mrocklin) [reply geopandas/tests/test_geoseries.py]: Reverted.  I think that this test was changed for a previous commit in this PR that has since been overwritten.

File: geopandas/geoseries.py (hunk 8 above)
Comment (jorisvandenbossche) [reply geopandas/geoseries.py]: Yes, I am fine with not including those kwargs

(it is there in pandas mainly for dataframe/series compatibility, as many functions are implemented for both with common code, although that is not the case for dropna)

**Comment by jorisvandenbossche** [comment]
@mrocklin go ahead with merging both prs! I will let you do it so you can merge the one first you prefer, as they will conflict with each other

**Comment by jorisvandenbossche** [comment]
@mrocklin Thanks!
